src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/AvoidDoubleBraceInitializationCheck.java

```diff
@@ -0,0 +1,141 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2020 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.coding;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+
+import com.puppycrawl.tools.checkstyle.StatelessCheck;
+import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtil;
+
+/**
+ * <p>
+ * Detects double brace initialization.
+ * </p>
+ * <p>
+ * Rationale: Double brace initialization may look cool, but it is considered
+ * as anti-pattern and should be avoided.
+ * This is also can lead to a hard-to-detect memory leaks, since created anonymous class
+ * is not static, it holds an implicit reference to the outer class instance.
+ * See this
+ * <a href="https://blog.jooq.org/2014/12/08/dont-be-clever-the-double-curly-braces-anti-pattern/">
+ * blog post</a> and
+ * <a href="https://www.baeldung.com/java-double-brace-initialization">
+ * article</a> for more details.
+ * </p>
+ * <p>
+ * To configure the check:
+ * </p>
+ * <pre>
+ * &lt;module name=&quot;AvoidDoubleBraceInitialization&quot;/&gt;
+ * </pre>
+ * <p>
+ * Which results in the following violations:
+ * </p>
+ * <pre>
+ * class MyClass {
+ *   List list1 = new ArrayList&lt;Object&gt; { // violation
+ *     {
+ *       add(new Object());
+ *     }
+ *   };
+ *   List list2 = new ArrayList&lt;Object&gt; { // violation, comments are alone commas are ignored
+ *     ;
+ *     // my comment
+ *     {
+ *       add(new Object());
+ *     }
+ *   };
+ *   List list3 = new ArrayList&lt;Object&gt; { // OK, there is an additional field
+ *     private int field;
+ *     {
```

Comment:
Why does the field matter? Isn't there still a concern for a memory leak?

Reply:
It is, but other class members can produce side effects, which can cause false positives.
So this check is focused only on "pure" double-brace init.

Reply:
Then shouldn't we add some warning or message saying so?

Reply:
Can you give an example of the message? I am not really understand your point.

Double brace initialization means anon class when its body contains only init blocks. In this case this anon instance has not only initialization, but something else.

Reply:
Maybe something like in the description.

This check strictly only looks for double brace initialization and therefore will ignore certain cases where the anonymous class can have fields or methods. (Insert Example) Though these will create the same memory issues as double brace, the extra fields/methods can produce side effects (Insert some reason/example if possible) if changed incorrectly.

Reply:
changed

Reply:
> the extra class members can produce side effects if changed incorrectly.

I do not understand this phrase. @rnveach, can you explain what you mean by `changed incorrectly` ?

Reply:
`Check is detecting only simple cases.`
As for me, this is a duplication of the first sentence.
`Check only looks for double brace initialization and it ignores cases
          where the anonymous class has fields or methods.`

Reply:
sounds good to me.

@rnveach , please review.

Reply:
> the extra class members can produce side effects if changed incorrectly.

This part of my sentence was in relation to https://github.com/checkstyle/checkstyle/pull/7457#discussion_r371770798 where you said
> It is, but other class members can produce side effects, which can cause false positives.

I took false positives to mean they could be changed but had to be done carefully. Feel free to add your own verbage mine was just an example.


I am mostly just looking for something that says "We check this type. This other type still has the same issue, but we can't flag it for XYZ. Here is an example of this other type."

Reply:
@strkkk , please show cases of real code where fields/methods are used, you probably found them during regression. Please share them.

 https://rules.sonarsource.com/java/RSPEC-3599 , there is no mentioning of field/method. It does not mean that such spec is right, but it is a point.
https://pmd.github.io/latest/pmd_rules_java_bestpractices.html#doublebraceinitialization - here too.
What about summary:
`Detects anonymous class which only has a class cody as single instance initializer.`
After Rationale (but before examples) we can make note: `This Check ignores any comments and semicolons in class body.`
Doublebrace is human term because it usually looks like `{{` but it is not a special a part of language.
Instance Initializers - https://docs.oracle.com/javase/specs/jls/se12/html/jls-8.html#jls-8.6
class body keep initializers - https://docs.oracle.com/javase/specs/jls/se12/html/jls-8.html#jls-8.1.6
So name of Check can be in human friendly as how usually users know this, but lets be JDK terminology specific in documentation to be exact.

By the way .... we need tests for few initializers in class body.

Reply:
@romani I don't understand your point.
1. I do not have real code with fields/methods (where can I take them from?)
2. There can be several initializers and there is a test case for it (for 2 init blocks). 
3. Do you mean to change implementation to detect only one init block? In my mind, several blocks or one - doesn't matter, it is same issue.
4. 
> but lets be JDK terminology specific in documentation to be exact

What do you mean? I think it is make sense to use common "doublebrace"

Reply:
> I do not have real code with fields/methods (where can I take them from?)

in such cases the most easy approach is to modify Check to report only cases with fields and methods and run regression. 
But looks like we do not need this as we now almost agree that such case is not boublebrace pattern at all, it is just fancy annonymous class with methods/fields and initializer.

> There can be several initializers and there is a test case for it (for 2 init blocks)

ok, I see them now, my bad.
It would be good to add case `new ArrayList<Object>() {{ add("1");}{ add("2");}};`.
as it is most used form. You have cases where `{` are on different lines, event is same, better to have case in form that is widely used (braces close to each other - `{{`).

> Do you mean to change implementation to detect only one init block? In my mind, several blocks or one - doesn't matter, it is same issue.

No problem, to report on several too, we just need to adjust description to be clear on what we detect - `Detects anonymous class which only has a class cody as set of instance initializers.`

> What do you mean?

I meant to extend description ..... `Rationale: Double brace initialization (set of <a href="https://docs.oracle.com/javase/specs/jls/se12/html/jls-8.html#jls-8.6"> Instance Initializers</a> in class body) may look cool ... `

Reply:
I think `// OK, there is an additional field`
should be changed to "// OK as it is not pure double brace pattern"

Reply:
I am satisfied.

@rnveach , please finalize review.